content/blog/_posts/2021-04-rust-custom-test-harness.md

```diff
@@ -0,0 +1,296 @@
+  * libtest returns the pass/fail status of the execution
+
+
+### What do I need from integration testing?
+
+There isn’t anything libtest specifically offers to support integration testing patterns.
+
+Setup of a standard test environment, especially in a complex system, is essential for managing expected behavior when making code changes.
+
+Libtest does not assist with setup or teardown. I need the ability to abstract away the setup and teardown of my test environment from test code. 
+
+This task will be performed either way. Without harness support, setup/teardown will be performed via external shell scripts or padding the setup/teardown process within every single integration test...
+
+It isn’t convenient to manage setup and teardown in a different context than the integration test. This kind of testing overhead leads to hard to reproduce and time consuming mistakes.
+
+### Where do we get started with a custom test harness?
+
+We’re going to build all our integration tests into a single crate. This is recommended in order to speed up compile time ([[1](https://endler.dev/2020/rust-compile-times/#combine-all-integration-tests-in-a-single-binary)], [[2](https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html)])
+
+This is distinctive, because libtest will compile each of your `#[test]` labeled functions into its own binary crate (with its own `main()`) and executes it as part of the test. 
+
+First we’re going to create an integration test directory at the root of the crate where we’re going to build our integration test focused binary.
+
+```shell
+$ mkdir integration
+$ touch integration/main.rs
+
+# Then create a main() function in main.rs
+```
+
+In your Cargo.toml, you want to add 
+
+```toml
+# Cargo.toml
+
+[dev-dependencies]
+inventory = "0.1"
+
+[[test]]
+name = “integration”
+path = “integration/main.rs”
+harness = false
+```
+
+What this does is tell cargo test to not use libtest when running the `integration` test.
+
+Then when we run `cargo test integration`, what cargo will do is compile `integration/main.rs` and execute it in the same manner as `cargo run`. This is all a harness is from `cargo`’s perspective. 
+
+(We'll revisit the `inventory` crate later in the post)
+
+
+### Add Setup and teardown steps
+
+First thing we’ll do is lay the foundation for our testing pattern. We’ll create 2 functions, `setup()` and `teardown()`, and add them to our `main()` (with reserved space inbetween for our future tests to be called).
+
+```rust
+// main.rs
+
+fn setup() {
+   println!("Setup")
+}
+
+fn teardown() {
+   println!("Teardown")
+}
+fn main() {
+   // Setup test environment
+   setup();
+
+   // TODO: Run the test
+
+   // Teardown test environment
+   teardown();
+}
+```
+
+### Collect all integration tests
+
+What do I mean? Just use `#[test]`, right? That’s what I was thinking at least. Turns out that is you opt out of using `#[test]` with `harness = false`. 
```

Comment:
I think this could be worded better, or at least use some extra context. Maybe something like:

> The next job our test runner needs to do is to create a list of all the functions annotated as tests. I thought there would be an easy way to do this by leveraging the `#[test]` attribute, but it turns out that there is no straightforward way to do this.

If you go with something more along these lines, you may want to just blend it into the next paragraph about libtest